### PR TITLE
Move `PyErr_NormalizeException()` up a few lines

### PR DESCRIPTION
**Pull request by rwgk**
## Description
The only production code change is to move `PyErr_NormalizeException()` up a few lines in `detail::error_string()`.

The 1st commit (e564142bb5e2c324e581a7bac8371f0c0ff5dc36) purely adds test coverage, without making changes to the production code. This effectively documents the current `error_already_set::what()` behavior.

The 2nd commit (254bfeb2a0f906b523c62825999d1ab8a2a4da7c) moves `PyErr_NormalizeException()` up a few lines in `detail::error_string()` and adjusts the new tests accordingly, effectively documenting the behavior change, which can be summarized as:

* It fixes the "happy" tests.
* The behavior for the two corner cases has more integrity but is still not really great. This was already addressed under PR #1895.

Note that all existing tests pass as-is! Only the newly added tests need adjustments.

Background: The production code change was adopted from PR #3964. The new tests are based on similar code under PR #1895. PR #1895 has several behavior changes. This PR handles one behavior change in isolation, for easier review and a clearer development history.
 

## Suggested changelog entry:

```rst
`error_already_set::what()` now handles non-normalized exceptions correctly.
```

### CODE REVIEW

File: tests/test_exceptions.cpp [hunk 1]
```diff
@@ -299,4 +299,12 @@ TEST_SUBMODULE(exceptions, m) {
             std::throw_with_nested(std::runtime_error("Outer Exception"));
         }
     });
+
+    m.def("error_already_set_what", [](const py::object &exc_type, const py::object &exc_value) {
+        PyErr_SetObject(exc_type.ptr(), exc_value.ptr());
+        std::string what = py::error_already_set().what();
+        bool py_err_set_after_what = (PyErr_Occurred() != nullptr);
+        PyErr_Clear();
+        return py::make_tuple(what, py_err_set_after_what);
```
Comment (Skylion007) [inline tests/test_exceptions.cpp:308]:
```suggestion
        return py::make_tuple(std::move(what), py_err_set_after_what);
```

File: tests/test_exceptions.py [hunk 2]
```diff
@@ -300,6 +300,7 @@ def test_error_already_set_what_with_happy_exceptions(
     assert what == expected_what
 
 
+@pytest.mark.skipif("env.PYPY", reason="PyErr_NormalizeException Segmentation fault")
```
Comment (Skylion007) [inline tests/test_exceptions.py:303]: Does this still segfault at the original Normalize location?

File: include/pybind11/detail/type_caster_base.h [hunk 3]
```diff
@@ -486,9 +488,6 @@ PYBIND11_NOINLINE std::string error_string() {
     if (scope.value) {
         errorString += (std::string) str(scope.value);
     }
-
-    PyErr_NormalizeException(&scope.type, &scope.value, &scope.trace);
-
     if (scope.trace != nullptr) {
```
[the remark this answers is not in the document]
Comment (rwgk) [reply include/pybind11/detail/type_caster_base.h]: Good point. I missed that.

File: tests/test_exceptions.py (hunk 2 above)
Comment (rwgk) [reply tests/test_exceptions.py:303]: Yes: https://github.com/pybind/pybind11/actions/runs/2386632518

File: tests/test_exceptions.cpp (hunk 1 above)
Comment (rwgk) [reply tests/test_exceptions.cpp]: Done.

**Comment by rwgk** [comment]
Looks like a GHA infrastructure bug popped up in the few minutes between the last two CI runs.

I'll try again a little later, but the only thing that changed was the extra `std::move()`, which seems to work for all non-Windows CI jobs.

```
Run jwlawson/actions-setup-cmake@v1.12
  with:
    github-api-token: ***
    use-3[2](https://github.com/pybind/pybind11/runs/6599035794?check_suite_focus=true#step:6:2)bit: false
  env:
    PIP_ONLY_BINARY: numpy
    FORCE_COLOR: [3](https://github.com/pybind/pybind11/runs/6599035794?check_suite_focus=true#step:6:3)
    PYTEST_TIMEOUT: 300
    pythonLocation: C:\hostedtoolcache\windows\Python\3.9.12\x6[4](https://github.com/pybind/pybind11/runs/6599035794?check_suite_focus=true#step:6:4)
Error: Could not find win32 asset for cmake version 3.23.2
```

**Comment by rwgk** [comment]
Close-opening to re-trigger CI, hoping the GHA infrastructure failure was resolved in the meantime.

**Comment by rwgk** [comment]
@Skylion007 thanks for your suggestions, they are applied now. I also answered the question about the PyPy segfault (yes, even without the production code change). The CI is green again. Could you please take another look?

**Review by Skylion007 — APPROVED** [review]
@rwgk Is this still a draft PR?

**Comment by rwgk** [comment]
> @rwgk Is this still a draft PR?

No sorry, I keep forgetting to click the Ready for review button.

**Review by Skylion007 — APPROVED** [review]

**Comment by Skylion007** [comment]
LGTM, merge it.

**Comment by rwgk** [comment]
Thanks, done! I'll work on updating #1895 (I assume there will be merge conflicts).
